server/src/main/java/org/opensearch/cluster/decommission/DecommissionAttributeMetadata.java

```diff
@@ -0,0 +1,227 @@
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.decommission;
+
+import org.opensearch.OpenSearchParseException;
+import org.opensearch.Version;
+import org.opensearch.cluster.AbstractNamedDiffable;
+import org.opensearch.cluster.NamedDiff;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.metadata.Metadata.Custom;
+import org.opensearch.common.Strings;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.xcontent.ToXContent;
+import org.opensearch.common.xcontent.XContentBuilder;
+import org.opensearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.Objects;
+
+/**
+ * Contains metadata about decommission attribute
+ *
+ * @opensearch.internal
+ */
+public class DecommissionAttributeMetadata extends AbstractNamedDiffable<Custom> implements Custom {
+
+    public static final String TYPE = "decommissionedAttribute";
+
+    private final DecommissionAttribute decommissionAttribute;
+    private final DecommissionStatus status;
+    public static final String attributeType = "awareness";
+
+    /**
+     * Constructs new decommission attribute metadata with given status
+     *
+     * @param decommissionAttribute attribute details
+     * @param status                current status of the attribute decommission
+     */
+    public DecommissionAttributeMetadata(DecommissionAttribute decommissionAttribute, DecommissionStatus status) {
+        this.decommissionAttribute = decommissionAttribute;
+        this.status = status;
+    }
+
+    /**
+     * Constructs new decommission attribute metadata with status as {@link DecommissionStatus#DECOMMISSION_INIT}
+     *
+     * @param decommissionAttribute attribute details
+     */
+    public DecommissionAttributeMetadata(DecommissionAttribute decommissionAttribute) {
+        this(decommissionAttribute, DecommissionStatus.DECOMMISSION_INIT);
+    }
+
+    /**
+     * Returns the current decommissioned attribute
+     *
+     * @return decommissioned attributes
+     */
+    public DecommissionAttribute decommissionAttribute() {
+        return this.decommissionAttribute;
+    }
+
+    /**
+     * Returns the current status of the attribute decommission
+     *
+     * @return attribute type
+     */
+    public DecommissionStatus status() {
+        return this.status;
+    }
+
+    /**
+     * Creates a new instance that has the given decommission attribute moved to the given @{@link DecommissionStatus}
+     * @param status status to be updated with
+     * @return new instance with updated status
+     */
+    public DecommissionAttributeMetadata withUpdatedStatus(DecommissionStatus status) {
+        return new DecommissionAttributeMetadata(decommissionAttribute(), status);
+    }
```

Comment:
nit: Do we strictly want this to be immutable?

Reply:
No strong opinion. But this would help when there's a reference to the old object elsewhere. Hence avoiding to 'reuse' anything